qa/rpc-tests/wallet_listreceived.py

```diff
@@ -15,19 +22,171 @@
+        assert_equal(txid_shielding1, r[0]['txid'])
+        assert_equal(10, r[0]['amount'])
+        assert_equal(1000000000, r[0]['amountZat'])
+        assert_false(r[0]['change'], "Note should not be change")
+        assert_equal(no_memo, r[0]['memo'])
+        assert_equal(0, r[0]['confirmations'])
+        assert_equal(-1, r[0]['blockindex'])
+        assert_equal(0, r[0]['blockheight'])
+
+        c = self.nodes[1].z_getnotescount(0)
+        assert_equal(1, c['sprout'], "Count of unconfirmed notes should be 1")
+
+        # Confirm transaction (10 ZEC shielded)
+        self.generate_and_sync(height+3)
+
+        # adjust confirmations
+        r[0]['confirmations'] = 1
+        # adjust blockindex
+        r[0]['blockindex'] = 1
+        # adjust height
+        r[0]['blockheight'] = height + 3
+
+        # Require one confirmation, note should be present
+        assert_equal(r, self.nodes[1].z_listreceivedbyaddress(zaddr1))
+
+        taddr = self.nodes[1].getnewaddress()
+        # Generate some change by sending part of zaddr1 back to taddr
+        opid = self.nodes[1].z_sendmany(zaddr1, [{'address': taddr, 'amount': 0.6}])
+        txid = wait_and_assert_operationid_status(self.nodes[1], opid)
+
+        self.generate_and_sync(height+4)
+
+        # Decrypted transaction details should be correct
+        pt = self.nodes[1].z_viewtransaction(txid)
+        assert_equal(pt['txid'], txid)
+        assert_equal(len(pt['spends']), 1)
+        # TODO: enable once z_viewtransaction displays transparent elements
+        # assert_equal(len(pt['outputs']), 2)
+        assert_equal(len(pt['outputs']), 1)
+
+        assert_equal(pt['spends'][0]['type'], 'sprout')
+        assert_equal(pt['spends'][0]['txidPrev'], txid_shielding1)
+        assert_equal(pt['spends'][0]['js'], 0)
+        assert_equal(pt['spends'][0]['jsPrev'], 0)
+        assert_equal(pt['spends'][0]['jsOutputPrev'], jsOutputPrev)
+        assert_equal(pt['spends'][0]['address'], zaddr1)
+        assert_equal(pt['spends'][0]['value'], Decimal('10.0'))
+        assert_equal(pt['spends'][0]['valueZat'], 1000000000)
+
+        # We expect a transparent output and a Sprout output. Output order can
+        # be randomized.
+        outputs.append({
+            'type': pt['outputs'][0]['type'],
+            'address': pt['outputs'][0]['address'],
+            'value': pt['outputs'][0]['value'],
+            'valueZat': pt['outputs'][0]['valueZat'],
+            'memo': pt['outputs'][0]['memo'],
+        })
+
+        # TODO: enable once z_viewtransaction displays transparent elements
+        #
+        # outputs.append({
+        #     'type': ['outputs'][1]['type'],
+        #     'address': pt['outputs'][1]['address'],
+        #     'value': pt['outputs'][1]['value'],
+        #     'valueZat': pt['outputs'][1]['valueZat'],
+        #     'memo': pt['outputs'][1]['memo'],
+        # })
+        #
```

Comment:
Change this to be resilient against the RPC returning its results in either order (since the transparent outputs wouldn't have `memo` fields):
```suggestion
        outputs = [{
            'type': output['type'],
            'address': output['address'],
            'value': output['value'],
            'valueZat': output['valueZat'],
        } for output in pt['outputs']]
        for (i, output) in enumerate(pt):
            if 'memo' in output:
                outputs[i]['memo'] = output['memo']

        # TODO: enable once z_viewtransaction displays transparent elements
```